### PR TITLE
387 strict node and edge idxs

### LINKED ISSUE #387: Strong typing for sDAG node id and edge index
Currently both node id and edge index in SubsplitDAG are of type size_t. We can use the named type pattern to give them zero-cost strong typing and enforce correct usage via the compiler.

### PR DESCRIPTION
**Pull request by davidrich27**
## Description

- Adds template for strongly typed datatypes, which form a lightweight wrapper for primitive types. 
- Enforces explicit casting to/from underlying primitive and other strong types.
- Template inherit hashes from primitives for stl maps.
- Introduces strong types for `NodeId`, `EdgeId`, and `TaxonId` from `SubsplitDAG`.  

Closes #387

## Tests

N/A: No additional testing is required.

## Checklist:

* [x] Code follows our detailed [contribution guidelines](CONTRIBUTING.md)
* [x] `clang-format` has been run
* [x] TODOs have been eliminated from the code
* [x] Comments are up to date, document intent, and there are no commented-out code blocks
* [ ] Issue branch has been squashed and rebased on main branch
* [ ] GitHub CI build on PR branch completed successfully

### CODE REVIEW

File: src/wrapper_sugar.hpp [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    os << (include_prefix ? PrefixToString() : "")
+       << "::" << ((value_ == NoId) ? "NoId" : std::to_string(value_));
+    return os.str();
+  }
+  friend std::ostream &operator<<(std::ostream &os, const SelfType &obj) {
+    os << obj.ToString(true);
+    return os;
+  }
+
+  UnderlyingType value_;
```
Comment (ognian-) [inline src/wrapper_sugar.hpp:82]: We could initialize to `NoId` here.

Comment (davidrich27) [reply src/wrapper_sugar.hpp]: Added initialized `NoId`.

File: src/sbn_maps.cpp [hunk 2]
```diff
@@ -117,6 +117,40 @@ IndexerBundle SBNMaps::BuildIndexerBundle(const BitsetSizeDict& rootsplit_counte
   return {rootsplits, indexer, index_to_child, parent_to_range, index};
 }
 
+IndexerBundleForDAG SBNMaps::BuildIndexerBundleForDAG(
```
[the remark this answers is not in the document]
Comment (davidrich27) [reply src/sbn_maps.cpp]: Removed codeblock and added typecast to DAG function.

**Review by matsen — APPROVED** [review]
Great!

**Pull request merged**